build_executable.jl windows support

I am trying to build a standalone executable on windows and I'm running into some problems.
I have added gcc as stated in the script.

First of I ran into a problem with a missing string escape when windows paths is written to a userimg file in `emit_userimgjl`. It causes the userimg to contain an `include("folder\folder\file.jl")` with only single `\`. By inserting include(\"$(escape_string(script_file))\") on line: https://github.com/JuliaLang/julia/blob/26ae20698cf0fc57ae6fc60ba42140884e71f558/contrib/build_executable.jl#L224 the path to the script is correct on windows.

After overcoming this problem I get this error:
```
INFO: Linking sys.dll
INFO: System image successfully built at C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin\libeps_to_pdf.dll
WARNING: Building sys.dll on Windows against LLVM < 3.5.0 can cause incorrect backtraces! Delete generated sys.dll to avoid these
problems
INFO: To run Julia with this image loaded, run: julia -J C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin\libeps_to_pdf.ji

running: gcc -g -Wl,--no-as-needed -IC:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia -IC:\Users\Hoegh\src -IC:\Users\Hoegh\src/
upport -IC:\Users\Hoegh\usr/include C:\Users\Hoegh\AppData\Local\Temp\jul9D3B.tmp\start_func.c -o C:\Users\Hoegh\Julia-0.4.0-dev_
9_1\bin\eps_to_pdf.exe -Wl,-rpath,C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin -LC:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin -ljulia -leps
to_pdf
ERROR: LoadError: could not spawn `gcc -g -Wl,--no-as-needed '-IC:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia' '-IC:\Users\Ho
gh\src' '-IC:\Users\Hoegh\src/support' '-IC:\Users\Hoegh\usr/include' 'C:\Users\Hoegh\AppData\Local\Temp\jul9D3B.tmp\start_func.c
 -o 'C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin\eps_to_pdf.exe' '-Wl,-rpath,C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin' '-LC:\Users\Hoe
h\Julia-0.4.0-dev_19_1\bin' -ljulia -leps_to_pdf`: no such file or directory (ENOENT)
 in _jl_spawn at process.jl:216 (repeats 2 times)
while loading C:\Users\Hoegh\Julia-0.4.0-dev_19_1\share\julia\build_executable.jl, in expression starting on line 247
```
When I run the command manually i get:
```
C:\Users\Hoegh\Julia-0.4.0-dev_19_1>gcc -g -Wl,--no-as-needed -IC:\Users\Hoegh\Julia-0.4.0-dev_19_
1\include\julia -IC:\Users\Hoegh\src -IC:\Users\Hoegh\src/support -IC:\Users\Hoegh\usr/include C:\
Users\Hoegh\AppData\Local\Temp\jul9D3B.tmp\start_func.c -o C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin
\eps_to_pdf.exe -Wl,-rpath,C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin -LC:\Users\Hoegh\Julia-0.4.0-de
v_19_1\bin -ljulia -leps_to_pdf
In file included from C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/uv.h:59:0,
                 from C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/ios.h:5,
                 from C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/libsupport.h:12,
                 from C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/julia.h:10,
                 from C:\Users\Hoegh\AppData\Local\Temp\jul9D3B.tmp\start_func.c:1:
C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/uv-win.h:241:3: error: unknown type name 'SRWLOCK'
```
The file I try to compile looks like:
```
function main()
    println("test")
end
```
@few do you have any idea what's wrong?

--- a/contrib/build_executable.jl
+++ b/contrib/build_executable.jl
@@ -110,7 +110,7 @@ function build_executable(exename, script_file, targetdir=nothing; force=false)
 
     gcc = find_system_gcc()
     # This argument is needed for the gcc, see issue #9973
-    win_arg = @windows ? "-D_WIN32_WINNT=0x0502" : ""
+    win_arg = @windows ? `-D_WIN32_WINNT=0x0502` : ``
     incs = get_includes()
     ENV2 = deepcopy(ENV)
     @windows_only begin
@@ -197,9 +197,9 @@ function emit_cmain(cfile, exename, relocation)
     if relocation
         sysji = joinpath("lib"*exename)
     else
-        sysji = joinpath("..", "lib", "lib"*exename)
+        sysji = joinpath(dirname(Sys.dlpath("libjulia")), "lib"*exename)
     end
-
+    sysji = escape_string(sysji)
     f = open(cfile, "w")
     write( f, """
         #include <julia.h>

--- a/contrib/build_executable.jl
+++ b/contrib/build_executable.jl
@@ -57,10 +57,9 @@ function build_executable(exename, script_file, targetdir=nothing; force=false)
     julia = abspath(joinpath(JULIA_HOME, "julia"))
     build_sysimg = abspath(joinpath(dirname(@__FILE__), "build_sysimg.jl"))
 
-    patchelf = nothing
     if targetdir != nothing
         patchelf = find_patchelf()
-        if patchelf == nothing
+        if patchelf == nothing && !(OS_NAME == :Windows)
             println("ERROR: Using the 'targetdir' option requires the 'patchelf' utility. Please install it.")
             return 1
         end
@@ -109,37 +108,62 @@ function build_executable(exename, script_file, targetdir=nothing; force=false)
     run(`$(julia) $(build_sysimg) $(sys.buildfile) native $(userimgjl) --force`)
     println()
 
+    gcc = find_system_gcc()
+    # This argument is needed for the gcc, see issue #9973
+    win_arg = @windows ? "-D_WIN32_WINNT=0x0600" : ""
+    @windowsxp_only win_arg = "-D_WINVER=0x0502"
+    incs = get_includes()
+    @windows_only begin
+        if contains(gcc, "WinRPM")
+            # This should not bee necessary, it is done due to WinRPM's gcc's
+            # include paths is not correct see WinRPM.jl issue #38
+            old_path = ENV["path"]
+            ENV["path"] = old_path*";"*dirname(gcc)
+            push!(incs, "-I"*abspath(joinpath(dirname(gcc),"..","include")))
+        else
+            gcc = Base.shell_split(gcc)
+        end
+    end
+
     for exe in [exe_release, exe_debug]
-        incs = join(get_includes(), " ")
-        println("running: gcc -g -Wl,--no-as-needed $(incs) $(cfile) -o $(exe.buildfile) -Wl,-rpath,$(sys.buildpath) -L$(sys.buildpath) $(exe.libjulia) -l$(exename)")
-        run(`gcc -g -Wl,--no-as-needed $(get_includes()) $(cfile) -o $(exe.buildfile) -Wl,-rpath,$(sys.buildpath) -L$(sys.buildpath) $(exe.libjulia) -l$(exename)`)
+        println("running: $gcc -g -Wl,--no-as-needed $win_arg $(join(incs, " ")) $(cfile) -o $(exe.buildfile) -Wl,-rpath,$(sys.buildpath) -L$(sys.buildpath) $(exe.libjulia) -l$(exename)")
+        run(`$gcc -g -Wl,--no-as-needed $win_arg $(incs) $(cfile) -o $(exe.buildfile) -Wl,-rpath,$(sys.buildpath) -L$(sys.buildpath) $(exe.libjulia) -l$(exename)`)
         println()
+    end
+    @windows_only begin
+        if contains(gcc, "WinRPM")
+            ENV["path"] = old_path
+        end
     end
 
     println("running: rm -rf $(tmpdir) $(sys.buildfile).o $(sys.buildfile0).o $(sys.buildfile0).ji")
-    run(`rm -rf $(tmpdir) $(sys.buildfile).o $(sys.buildfile0).o $(sys.buildfile0).ji`)
+    map(f-> rm(f, recursive=true), [tmpdir, sys.buildfile*".o", sys.buildfile0*".o", sys.buildfile0*".ji"])
     println()
 
     if targetdir != nothing
         # Move created files to target directory
-        run(`mv $(exe_release.buildfile) $(exe_debug.buildfile) $(sys.buildfile).$(Sys.dlext) $(sys.buildfile).ji $(targetdir)`)
+        for file in [exe_release.buildfile, exe_debug.buildfile, sys.buildfile*".$(Sys.dlext)", sys.buildfile*".ji"]
+            mv(file,joinpath(targetdir, basename(file)))
+        end
 
         # Copy needed shared libraries to the target directory
         tmp = ".*\.$(Sys.dlext).*"
         shlibs = filter(Regex(tmp),readdir(sys.buildpath))
         for shlib in shlibs
-            run(`cp $(joinpath(sys.buildpath, shlib)) $(targetdir)`)
+            cp(joinpath(sys.buildpath, shlib), joinpath(targetdir, shlib))
         end
 
-        # Fix rpath in executable and shared libraries
-        shlibs = filter(Regex(tmp),readdir(targetdir))
-        push!(shlibs, exe_release.filename)
-        push!(shlibs, exe_debug.filename)
-
-        for shlib in shlibs
-            rpath = readall(`$(patchelf) --print-rpath $(joinpath(targetdir, shlib))`)[1:end-1]
-            if Base.samefile(rpath, sys.buildpath)
-                run(`$(patchelf) --set-rpath $(targetdir) $(joinpath(targetdir, shlib))`)
+        @unix_only begin
+            # Fix rpath in executable and shared libraries
+            shlibs = filter(Regex(tmp),readdir(targetdir))
+            push!(shlibs, exe_release.filename)
+            push!(shlibs, exe_debug.filename)
+            println(shlibs)
+            for shlib in shlibs
+                rpath = readall(`$(patchelf) --print-rpath $(joinpath(targetdir, shlib))`)[1:end-1]
+                if Base.samefile(rpath, sys.buildpath)
+                    run(`$(patchelf) --set-rpath $(targetdir) $(joinpath(targetdir, shlib))`)
+                end
             end
         end
     end
@@ -151,20 +175,13 @@ function build_executable(exename, script_file, targetdir=nothing; force=false)
 end
 
 function find_patchelf()
-    patchelf = nothing
-
-    for x in [joinpath(JULIA_HOME, "patchelf"), "patchelf"]
-        patchelf = x
-        works = true
+    for patchelf in [joinpath(JULIA_HOME, "patchelf"), "patchelf"]
         try
-            x=readall(`$(patchelf) --version`)
-        catch
-            works = false
+            if success(`$(patchelf) --version`)
+                return patchelf
+            end
         end
-        works && break
     end
-
-    patchelf
 end
 
 function get_includes()
@@ -219,12 +236,30 @@ function emit_cmain(cfile, exename, relocation)
 end
 
 function emit_userimgjl(userimgjl, script_file)
-    f = open(userimgjl, "w")
-    write( f, """
-        include(\"$(script_file)\")
-        """
-    )
-    close(f)
+    open(userimgjl, "w") do f
+        write( f, "include(\"$(escape_string(script_file))\")")
+    end
+end
+
+function find_system_gcc()
+    # On Windows, check to see if WinRPM is installed, and if so, see if gcc is installed
+    @windows_only try
+        require("WinRPM")
+        winrpmgcc = joinpath(WinRPM.installdir,"usr","$(Sys.ARCH)-w64-mingw32",
+            "sys-root","mingw","bin","gcc.exe")
+        if success(`$winrpmgcc --version`)
+            return winrpmgcc
+        end
+    end
+
+    # See if `gcc` exists
+    try
+        if success(@windows ? `cmd /c gcc -v` : `gcc -v`)
+            return @windows ? "cmd /c gcc" : "gcc"
+        end
+    end
+
+    error( "GCC not found on system: " * @windows? "GCC can be installed via `Pkg.add(\"WinRPM\"); WinRPM.install(\"gcc\")`" : "" )
 end
 
 if !isinteractive()

--- a/contrib/build_executable.jl
+++ b/contrib/build_executable.jl
@@ -110,30 +110,23 @@ function build_executable(exename, script_file, targetdir=nothing; force=false)
 
     gcc = find_system_gcc()
     # This argument is needed for the gcc, see issue #9973
-    win_arg = @windows ? "-D_WIN32_WINNT=0x0600" : ""
-    @windowsxp_only win_arg = "-D_WINVER=0x0502"
+    win_arg = @windows ? "-D_WIN32_WINNT=0x0502" : ""
     incs = get_includes()
+    ENV2 = deepcopy(ENV)
     @windows_only begin
         if contains(gcc, "WinRPM")
             # This should not bee necessary, it is done due to WinRPM's gcc's
             # include paths is not correct see WinRPM.jl issue #38
-            old_path = ENV["path"]
-            ENV["path"] = old_path*";"*dirname(gcc)
+            ENV2["PATH"] *= ";" * dirname(gcc)
             push!(incs, "-I"*abspath(joinpath(dirname(gcc),"..","include")))
-        else
-            gcc = Base.shell_split(gcc)
         end
     end
 
     for exe in [exe_release, exe_debug]
         println("running: $gcc -g -Wl,--no-as-needed $win_arg $(join(incs, " ")) $(cfile) -o $(exe.buildfile) -Wl,-rpath,$(sys.buildpath) -L$(sys.buildpath) $(exe.libjulia) -l$(exename)")
-        run(`$gcc -g -Wl,--no-as-needed $win_arg $(incs) $(cfile) -o $(exe.buildfile) -Wl,-rpath,$(sys.buildpath) -L$(sys.buildpath) $(exe.libjulia) -l$(exename)`)
+        cmd = setenv(`$gcc -g -Wl,--no-as-needed $win_arg $(incs) $(cfile) -o $(exe.buildfile) -Wl,-rpath,$(sys.buildpath) -L$(sys.buildpath) $(exe.libjulia) -l$(exename)`, ENV2)
+        run(cmd)
         println()
-    end
-    @windows_only begin
-        if contains(gcc, "WinRPM")
-            ENV["path"] = old_path
-        end
     end
 
     println("running: rm -rf $(tmpdir) $(sys.buildfile).o $(sys.buildfile0).o $(sys.buildfile0).ji")
@@ -253,9 +246,9 @@ function find_system_gcc()
     end
 
     # See if `gcc` exists
-    try
-        if success(@windows ? `cmd /c gcc -v` : `gcc -v`)
-            return @windows ? "cmd /c gcc" : "gcc"
+    @unix_only try
+        if success(`gcc -v`)
+            return "gcc"
         end
     end
 

